# Data Understanding

In [1]:
import pandas as pd
df = pd.read_csv('flight.csv')

df.sample(10)

MEMBER_NO    FFP_DATE FIRST_FLIGHT_DATE  GENDER  FFP_TIER  \
18894      42986   9/25/2012        10/15/2012    Male         4   
21447      40622   4/10/2010         4/10/2010    Male         4   
54249      57235    5/7/2008        12/22/2008  Female         4   
18232      57621  10/16/2008        12/17/2008    Male         4   
15225      37203   6/20/2007         7/27/2007    Male         4   
13180      57799    8/3/2009        10/18/2009    Male         4   
30670        934   9/12/2006          6/2/2011    Male         4   
52613       7454   3/15/2013         3/21/2013    Male         4   
32139      46075    1/6/2007         3/23/2007    Male         4   
33031      48274  10/10/2009        10/10/2009  Female         4   

           WORK_CITY  WORK_PROVINCE WORK_COUNTRY   AGE  LOAD_TIME  ...  \
18894        beijing        beijing           CN  39.0  3/31/2014  ...   
21447      guangzhou      guangdong           CN  36.0  3/31/2014  ...   
54249       shenzhen      guangdong           CN  34.0  3/31/2014  ...   
18232            NaN            NaN           SG  50.0  3/31/2014  ...   
15225      guangzhou      guangdong           CN  41.0  3/31/2014  ...   
13180         wuzhou        guangxi           CN  35.0  3/31/2014  ...   
30670       shanghai       shanghai           CN  53.0  3/31/2014  ...   
52613    shenyangshi  liaoningsheng           CN  57.0  3/31/2014  ...   
32139  Clayton South       Victoria           AU  33.0  3/31/2014  ...   
33031       changsha          hunan           CN  27.0  3/31/2014  ...   

       SUM_YR_2  SEG_KM_SUM  LAST_FLIGHT_DATE  LAST_TO_END  AVG_INTERVAL  \
18894    3315.0       13825         1/10/2014           82     50.222222   
21447       0.0       15751        11/25/2012          493     44.250000   
54249    1393.0        4125         6/22/2013          284    131.500000   
18232       0.0       18240        11/26/2012          492     37.000000   
15225   10709.0       22567         3/31/2014            1     22.687500   
13180   13201.0       25574         3/14/2014           18     35.388889   
30670    2125.0        9922         1/23/2014           69     64.000000   
52613    1648.0        4221         3/21/2014           11    183.000000   
32139       0.0        8801         6/24/2012          647      2.000000   
33031    1540.0       12172          7/2/2013          274     56.714286   

      MAX_INTERVAL  EXCHANGE_COUNT  avg_discount  Points_Sum  Point_NotFlight  
18894          107               0      0.936198       12323                0  
21447           93               0      0.723002        6226                0  
54249          212               0      0.514861        7516                2  
18232          123               0      0.736985        9074                0  
15225           88               2      0.697413       27115               39  
13180          168               0      0.694051       22307               16  
30670          236               0      0.730515        6378                0  
52613          273               0      0.560836        1842                0  
32139            2               0      0.768532        7538                0  
33031          256               0      0.533171        4532                0  

[10 rows x 23 columns]

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62988 entries, 0 to 62987
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   MEMBER_NO          62988 non-null  int64  
 1   FFP_DATE           62988 non-null  object 
 2   FIRST_FLIGHT_DATE  62988 non-null  object 
 3   GENDER             62985 non-null  object 
 4   FFP_TIER           62988 non-null  int64  
 5   WORK_CITY          60719 non-null  object 
 6   WORK_PROVINCE      59740 non-null  object 
 7   WORK_COUNTRY       62962 non-null  object 
 8   AGE                62568 non-null  float64
 9   LOAD_TIME          62988 non-null  object 
 10  FLIGHT_COUNT       62988 non-null  int64  
 11  BP_SUM             62988 non-null  int64  
 12  SUM_YR_1           62437 non-null  float64
 13  SUM_YR_2           62850 non-null  float64
 14  SEG_KM_SUM         62988 non-null  int64  
 15  LAST_FLIGHT_DATE   62988 non-null  object 
 16  LAST_TO_END        629

Dataset ini merupakan data pelanggan program *Frequent Flyer*, yang berisi informasi tentang perilaku dan karakteristik pelanggan. Berikut adalah penjelasan tentang dataset Ini:

Dataset ini berisi 62988 baris data dan 20 kolom, yang dapat dikategorikan menjadi beberapa bagian:

Informasi Identitas

- MEMBER_NO-b: ID unik untuk setiap pelanggan
- GENDER: Jenis kelamin pelanggan
- AGE: Umur pelanggan
- WORK_CITY, WORK_PROVINCE, WORK_COUNTRY: Informasi tentang lokasi asal pelanggan

Informasi Penerbangan

- FIRST_FLIGHT_DATE: Tanggal penerbangan pertama pelanggan
- LAST_FLIGHT_DATE: Tanggal penerbangan terakhir pelanggan
- FLIGHT_COUNT: Jumlah penerbangan yang telah dilakukan pelanggan
- SEG_KM_SUM: Total jarak (km) penerbangan yang telah dilakukan pelanggan

Informasi Program Frequent Flyer

- FFP_DATE: Tanggal bergabung dengan program Frequent Flyer
- FFP_TIER: Tier pelanggan dalam program Frequent Flyer
- BP_SUM: Rencana perjalanan pelanggan
- SUM_YR_1, SUM_YR_2: Total kredit/point di tahun pertama dan kedua
- Points_Sum: Jumlah poin yang telah didapat pelanggan
- Point_NotFlight: Poin yang tidak digunakan oleh pelanggan

Informasi Lain

- LOAD_TIME: Tanggal data diambil
- AVG_INTERVAL: Rata-rata jarak waktu antara penerbangan
- MAX_INTERVAL: Maksimal jarak waktu antara penerbangan
- EXCHANGE_COUNT: Jumlah penukaran
- avg_discount: Rata-rata diskon yang didapat pelanggan

# Preprocessing Data

## Handling Duplicated

In [4]:
print(df.duplicated().sum())

0


Tidak terdapat duplicated Data dari dataset ini

## Handling Missing Value

In [3]:
df.isnull().sum()

MEMBER_NO               0
FFP_DATE                0
FIRST_FLIGHT_DATE       0
GENDER                  3
FFP_TIER                0
WORK_CITY            2269
WORK_PROVINCE        3248
WORK_COUNTRY           26
AGE                   420
LOAD_TIME               0
FLIGHT_COUNT            0
BP_SUM                  0
SUM_YR_1              551
SUM_YR_2              138
SEG_KM_SUM              0
LAST_FLIGHT_DATE        0
LAST_TO_END             0
AVG_INTERVAL            0
MAX_INTERVAL            0
EXCHANGE_COUNT          0
avg_discount            0
Points_Sum              0
Point_NotFlight         0
dtype: int64

Output code di atas menunjukkan jumlah missing value yang ada di setiap kolom pada dataset. Berikut adalah interpretasi dari output tersebut:

- Kolom *MEMBER_NO, FFP_DATE, FFP_TIER, LOAD_TIME, FLIGHT_COUNT, BP_SUM, SEG_KM_SUM, LAST_FLIGHT_DATE, LAST_TO_END, AVG_INTERVAL, MAX_INTERVAL, EXCHANGE_COUNT, avg_discount, Points_Sum, dan Point_NotFlight* tidak memiliki missing value.
- Kolom GENDER memiliki 3 missing value, yang relatif sedikit dibandingkan dengan jumlah total data.
- Kolom WORK_CITY memiliki 2269 missing value, yang cukup signifikan dan mungkin perlu penanganan khusus.
- Kolom WORK_PROVINCE memiliki 3248 missing value, yang lebih banyak daripada WORK_CITY dan mungkin perlu penanganan khusus.
- Kolom WORK_COUNTRY memiliki 26 missing value, yang relatif sedikit.
- Kolom AGE memiliki 420 missing value, yang cukup signifikan dan mungkin perlu penanganan khusus.
- Kolom SUM_YR_1 memiliki 551 missing value, yang cukup signifikan dan mungkin perlu penanganan khusus.
- Kolom SUM_YR_2 memiliki 138 missing value, yang relatif sedikit.

Dalam keseluruhan, dataset ini memiliki beberapa kolom dengan missing value yang signifikan, terutama WORK_CITY, WORK_PROVINCE, dan AGE. Penanganan missing value ini perlu dilakukan sebelum melakukan analisis lebih lanjut untuk memastikan keakuratan dan keandalan hasil analisis.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62988 entries, 0 to 62987
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   MEMBER_NO          62988 non-null  int64  
 1   FFP_DATE           62988 non-null  object 
 2   FIRST_FLIGHT_DATE  62988 non-null  object 
 3   GENDER             62985 non-null  object 
 4   FFP_TIER           62988 non-null  int64  
 5   WORK_CITY          60719 non-null  object 
 6   WORK_PROVINCE      59740 non-null  object 
 7   WORK_COUNTRY       62962 non-null  object 
 8   AGE                62568 non-null  float64
 9   LOAD_TIME          62988 non-null  object 
 10  FLIGHT_COUNT       62988 non-null  int64  
 11  BP_SUM             62988 non-null  int64  
 12  SUM_YR_1           62437 non-null  float64
 13  SUM_YR_2           62850 non-null  float64
 14  SEG_KM_SUM         62988 non-null  int64  
 15  LAST_FLIGHT_DATE   62988 non-null  object 
 16  LAST_TO_END        629

In [5]:
df.describe()

MEMBER_NO      FFP_TIER           AGE  FLIGHT_COUNT         BP_SUM  \
count  62988.000000  62988.000000  62568.000000  62988.000000   62988.000000   
mean   31494.500000      4.102162     42.476346     11.839414   10925.081254   
std    18183.213715      0.373856      9.885915     14.049471   16339.486151   
min        1.000000      4.000000      6.000000      2.000000       0.000000   
25%    15747.750000      4.000000     35.000000      3.000000    2518.000000   
50%    31494.500000      4.000000     41.000000      7.000000    5700.000000   
75%    47241.250000      4.000000     48.000000     15.000000   12831.000000   
max    62988.000000      6.000000    110.000000    213.000000  505308.000000   

            SUM_YR_1       SUM_YR_2     SEG_KM_SUM   LAST_TO_END  \
count   62437.000000   62850.000000   62988.000000  62988.000000   
mean     5355.376064    5604.026014   17123.878691    176.120102   
std      8109.450147    8703.364247   20960.844623    183.822223   
min         0.000000       0.000000     368.000000      1.000000   
25%      1003.000000     780.000000    4747.000000     29.000000   
50%      2800.000000    2773.000000    9994.000000    108.000000   
75%      6574.000000    6845.750000   21271.250000    268.000000   
max    239560.000000  234188.000000  580717.000000    731.000000   

       AVG_INTERVAL  MAX_INTERVAL  EXCHANGE_COUNT  avg_discount   Points_Sum  \
count  62988.000000  62988.000000    62988.000000  62988.000000   62988.0000   
mean      67.749788    166.033895        0.319775      0.721558   12545.7771   
std       77.517866    123.397180        1.136004      0.185427   20507.8167   
min        0.000000      0.000000        0.000000      0.000000       0.0000   
25%       23.370370     79.000000        0.000000      0.611997    2775.0000   
50%       44.666667    143.000000        0.000000      0.711856    6328.5000   
75%       82.000000    228.000000        0.000000      0.809476   14302.5000   
max      728.000000    728.000000       46.000000      1.500000  985572.0000   

       Point_NotFlight  
count     62988.000000  
mean          2.728155  
std           7.364164  
min           0.000000  
25%           0.000000  
50%           0.000000  
75%           1.000000  
max         140.000000

In [7]:
df.describe(include='object')

FFP_DATE FIRST_FLIGHT_DATE GENDER  WORK_CITY WORK_PROVINCE  \
count       62988             62988  62985      60719         59740   
unique       3068              3406      2       3234          1165   
top     1/13/2011         2/16/2013   Male  guangzhou     guangdong   
freq          184                96  48134       9386         17509   

       WORK_COUNTRY  LOAD_TIME LAST_FLIGHT_DATE  
count         62962      62988            62988  
unique          118          1              731  
top              CN  3/31/2014        3/31/2014  
freq          57748      62988              959

In [18]:
# Handling kolom Gender dengan nilai modusnya
df['GENDER'].fillna(df['GENDER'].mode()[0], inplace=True)

C:\Users\ACER\AppData\Local\Temp\ipykernel_8772\2944669490.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['GENDER'].fillna(df['GENDER'].mode()[0], inplace=True)


In [ ]:
# Handling kolom lokasi dengan 'Unknown'
df[['WORK_CITY', 'WORK_PROVINCE']] = df[['WORK_CITY', 'WORK_PROVINCE']].fillna('Unknown')

In [ ]:
# Handling kolom country dengan nilai modusnya
df['WORK_COUNTRY'].fillna(df['WORK_COUNTRY'].mode().iloc[0], inplace=True)

In [ ]:
# Handling kolom AGE dengan nilai mediannya
df['AGE'].fillna(df['AGE'].median()[0], inplace=True)

In [17]:
df.isnull().sum()

MEMBER_NO              0
FFP_DATE               0
FIRST_FLIGHT_DATE      0
GENDER                 0
FFP_TIER               0
WORK_CITY              0
WORK_PROVINCE          0
WORK_COUNTRY           0
AGE                    0
LOAD_TIME              0
FLIGHT_COUNT           0
BP_SUM                 0
SUM_YR_1             551
SUM_YR_2             138
SEG_KM_SUM             0
LAST_FLIGHT_DATE       0
LAST_TO_END            0
AVG_INTERVAL           0
MAX_INTERVAL           0
EXCHANGE_COUNT         0
avg_discount           0
Points_Sum             0
Point_NotFlight        0
dtype: int64